rfcs/2021-05-05-7351-framing-and-codecs.md

```diff
@@ -0,0 +1,183 @@
+consist of _decoders_, which deserialize a byte frame into structured data, and
+_encoders_, which serialize structured data into a byte frame.
+
+Conceptually, we want reusable pieces of logic which would allow us to collapse
+a source + (decoder) transform into a source with decoder config, and to
+collapse an (encoder) transform + sink into a sink with encoder config, in a way
+that is transparent to the user. These transforms merely convert between
+encoding formats, e.g. from `bytes` to `json` and vice versa.
+
+## Scope
+
+The scope of this RFC concern at which level framing and decoding operates, how
+these framers and codecs can be configured, and how they can be shared in a
+uniform way. It does not cover any specific implementation for framing or a
+codec. Also see [future work](#future-work) for extended goals that are not
+covered by this RFC.
+
+## Motivation
+
+Currently, we have no explicit abstraction that is responsible for handling
+framing and codecs, such that each source and sink may implement them in their
+own way. This can introduce inconsistencies and additional maintenance burden.
+Components have their own defaults, making behavior unpredictable and surprising
+users, e.g. as documented in
+[#3453](https://github.com/timberio/vector/issues/3453).
+
+## Internal Proposal
+
+Since decoding and encoding should apply universally to all sources and sinks,
+it should be configurable by a shared field. This is accomplished by the
+`SourceOuter` and `SinkOuter` wrappers:
+
+```rust
+// Newly introduced wrapper, analogously to `SinkOuter`.
+struct SourceOuter {
+    framing: Framer,
+    decoding: Decoder,
+}
+```
+
+```rust
+struct SinkOuter {
+    // … existing fields
+    encoding: Encoder,
+}
+```
+
+Conceptually, decoding is applied _after_ a source has processed an event and
+encoding is applied _before_ a sink has processed an event. These can be
+implemented as `FunctionTransform`s in our pipeline, in the same way transforms
+are implemented.
+
+Framing takes place before an event is created since the message boundary
+required to form an event is not known yet. Therefore the source needs to be
+aware of framing and call the framer to determine message boundaries before
+creating an event.
+
+To make the source aware of framing, the `Framer` needs to be passed to the
+`SourceContext`, so that the source implementation can call it:
+
+```rust
+struct SourceContext {
+    // … existing fields
+    framing: Framer,
+}
+```
+
+In the context of Vector, decoders and encoders implement functions with the
+signature `fn decode(value: Value) -> Result<Value>` /
+`fn encode(value: Value) -> Result<Value>` where e.g. a `json` encoder can
+convert a `Value::Bytes` to `Value::Map`.
+
+For the implementation of a framer, we can defer to Tokio's
+[`FramedRead`](https://docs.rs/tokio-util/0.6.6/tokio_util/codec/struct.FramedRead.html)
+trait which does exactly what we want here.
```

Comment:
This isn't a complete solution because `FramedRead` still needs a `Decoder` instance with the actual logic for where to split messages. That trait is actually a bit of a combination of the ideas of framer and decoder here.

Reply:
This has been updated to use `FunctionTransform<Vec<u8>>` instead.